[QMR / GMRES / BICGSTAB] Support of complex numbers

platawiec opened the pull request:
This PR adds complex support for the following solvers:
* `qmr`
* `gmres`
* `bicgstab`

The intent is to show what it would take to start to address #103 , namely, the following for the remaining solvers:

* Separating the type of the history `T` (as in `SimpleStats{T}`) from the type parameter of the eltype the solver expects. This involves adding an additional solver type parameter
* Minor adjustments to the solver to respect solving over complex matrices.

As part of this PR, the type restrictions in `krylov_utils.jl` were relaxed, and meaningful tests for complex matrices were added.

amontoison commented:
Hi @platawiec !
Thank you for this pull request, I already start to adress #103 with the branch [`complex`](https://github.com/JuliaSmoothOptimizers/Krylov.jl/tree/complex) of `JSO/Krylov.jl`.
I updated headers, docstring and the number of parameters of krylov solvers (https://github.com/JuliaSmoothOptimizers/Krylov.jl/commit/af41e5ec50d188393a1adb0431cd5ea3aa047895).
I just rebased this branch with the last modifications.

Can you open your pull request to add your modifications to this branch ?

amontoison commented:
Is it also possible to open a pull request for each solver ?
It will be easier to review! :smiley:

platawiec commented:
I have changed the merge target to the `complex` branch and rebased onto the latest changes there.

If requested, I can open up separate PRs for each solver. However, the changes to the actual solvers were kept as minimal as possible and each one is a self-contained commit to aid in review. Most of the "lines changed" are from indenting the test set (since we just iterate over the testing datatypes). Let me know if you still wish to see separate PRs.

amontoison commented on src/bicgstab.jl:
```diff
@@ -41,6 +41,8 @@ Information will be displayed every `verbose` iterations.
 
 This implementation allows a left preconditioner `M` and a right preconditioner `N`.
 
+This solver support complex `A` and `b`.
```
```suggestion
This solver supports complex `A` and `b`.
```

amontoison commented on src/bicgstab.jl:
```diff
@@ -111,10 +113,10 @@ function bicgstab!(solver :: BicgstabSolver{T,FC,S}, A, b :: AbstractVector{FC};
   ε = atol + rtol * rNorm
-  (verbose > 0) && @printf("%5s  %7s  %8s  %8s\n", "k", "‖rₖ‖", "αₖ", "ωₖ")
-  display(iter, verbose) && @printf("%5d  %7.1e  %8.1e  %8.1e\n", iter, rNorm, α, ω)
+  (verbose > 0) && @printf("%5s  %7s  %8s  %8s\n", "k", "‖rₖ‖", "‖αₖ‖", "‖ωₖ‖")
```
```suggestion
  (verbose > 0) && @printf("%5s  %7s  %8s  %8s\n", "k", "‖rₖ‖", "|αₖ|", "|ωₖ|")
```

amontoison replied:
`αₖ` and `ωₖ` are scalars.

amontoison commented on src/gmres.jl:
```diff
@@ -29,6 +29,8 @@ This implementation allows a left preconditioner M and a right preconditioner N.
 - Right preconditioning : AN⁻¹u = b with x = N⁻¹u
 - Split preconditioning : M⁻¹AN⁻¹u = M⁻¹b with x = N⁻¹u
 
+This solver support complex `A` and `b`.
```
```suggestion
This solver supports complex `A` and `b`.
```

amontoison commented on src/krylov_utils.jl:
```diff
@@ -57,34 +57,37 @@ c real and (s, ρ) complexes such that
     [ s̅  -c ] [ b ] = [ 0 ].
 """
-function sym_givens(a :: Complex{T}, b :: Complex{T}) where T <: AbstractFloat
+function sym_givens(a, b)
```
Can you keep the original implementation and add the following methods ?
```
@inline sym_givens(a :: Complex{T}, b :: T) where T <: AbstractFloat = sym_givens(a, Complex{T}(b))
@inline sym_givens(a :: T, b :: Complex{T}) where T <: AbstractFloat = sym_givens(Complex{T}(a), b)
```

amontoison commented on src/qmr.jl:
```diff
@@ -105,14 +107,14 @@ function qmr!(solver :: QmrSolver{T,FC,S}, A, b :: AbstractVector{FC}; c :: Abst
-  sₖ₋₂ = sₖ₋₁ = sₖ = zero(T)  # Givens sines used for the QR factorization of Tₖ₊₁.ₖ
-  wₖ₋₂ .= zero(T)             # Column k-2 of Wₖ = Vₖ(Rₖ)⁻¹
-  wₖ₋₁ .= zero(T)             # Column k-1 of Wₖ = Vₖ(Rₖ)⁻¹
+  cₖ₋₂ = cₖ₋₁ = cₖ = zero(FC) # Givens cosines used for the QR factorization of Tₖ₊₁.ₖ
```
```suggestion
  cₖ₋₂ = cₖ₋₁ = cₖ = zero(T)  # Givens cosines used for the QR factorization of Tₖ₊₁.ₖ
```

amontoison commented on src/qmr.jl:
```diff
@@ -133,15 +135,15 @@ function qmr!(solver :: QmrSolver{T,FC,S}, A, b :: AbstractVector{FC}; c :: Abst
     mul!(p, Aᵀ, uₖ)  # Forms uₖ₊₁ : p ← Aᵀuₖ
 
-    @kaxpy!(n, -γₖ, vₖ₋₁, q)  # q ← q - γₖ * vₖ₋₁
+    @kaxpy!(n, -conj(γₖ), vₖ₋₁, q)  # q ← q - γₖ * vₖ₋₁
```
Can we avoid to change `γₖ` and `αₖ` into `conj(γₖ)` and `conj(αₖ)` if we permute the order of some dot products ?

amontoison replied:
I updated BiLQ (#497) for complex numbers and I confirm that we can avoid `conj(γₖ)` and `conj(αₖ)`.

amontoison replied:
```julia
"""
    β, γ, T, U, V = biortho(A, b, c, l)

AVₖ  = VₖTₖ    + βₖ₊₁vₖ₊₁(eₖ)ᵀ = Vₖ₊₁Tₖ₊₁.ₖ
AᵀUₖ = Uₖ(Tₖ)ᵀ + γ̄ₖ₊₁uₖ₊₁(eₖ)ᵀ = Uₖ₊₁(Tₖ.ₖ₊₁)ᵀ
(Vₖ)ᵀUₖ = (Uₖ)ᵀVₖ = Iₖ
"""
function biortho(A::Matrix{S}, b::Vector{S}, c::Vector{S}, l::Int) where S
  n, m = size(A)
  Aᵀ = A'
  T = spzeros(S, l+1, l+1)
  V = zeros(S, m, l+1)
  U = zeros(S, n, l+1)
  cᵗb = dot(c, b)
  β = √(abs(cᵗb))
  γ = cᵗb / β
  V[:,1] = b / β
  U[:,1] = c / conj(γ)
  for k = 1 : l
    q = A  * V[:, k]
    p = Aᵀ * U[:, k]
    αₖ = dot(U[:, k], q)
    T[k,k] = αₖ
    q = q -      αₖ  * V[:, k]
    p = p - conj(αₖ) * U[:, k]

    if k ≥ 2
      βₖ = T[k,k-1]
      γₖ = T[k-1,k]
      q = q - γₖ * V[:, k-1]
      p = p - βₖ * U[:, k-1]
    end

    pᵗq = dot(p, q)
    βₖ₊₁ = √(abs(pᵗq))
    γₖ₊₁ = pᵗq / βₖ₊₁
    V[:,k+1] = q / βₖ₊₁
    T[k+1,k] = βₖ₊₁

    U[:,k+1] = p / conj(γₖ₊₁)
    T[k,k+1] = γₖ₊₁
  end
  return β, γ, T, U, V
end

for T ∈ [Float64, ComplexF64]
  println("Type: $T")
  n = 10
  A = rand(T, n, n)
  b = rand(T, n)
  c = rand(T, n)
  l = 5
  p = 3
  β, γ, T, U, V = biortho(A, b, c, l)

  println(norm(A  * V[:,1:p] - V[:, 1:p+1] * T[1:p+1,1:p] ))
  println(norm(A' * U[:,1:p] - U[:, 1:p+1] * T[1:p,1:p+1]'))
  println(norm(U[:, 1:p]' * A  * V[:,1:p] - T[1:p,1:p] ))
  println(norm(V[:, 1:p]' * A' * U[:,1:p] - T[1:p,1:p]'))
  println(norm(U' * V - I))
  println(norm(V' * U - I))
  println()
end
```

amontoison commented:
@platawiec 
I did some basic benchmarks and it's more efficient to force the dispatch to BLAS in `krylov_utils.jl` instead of removing the type restrictions.
https://github.com/JuliaSmoothOptimizers/Krylov.jl/pull/489/files#diff-69fc705bdc794bf95e42c6cf19bb2de7c9379c29bd0cf42ba3442a106e9b0175

amontoison reviewed (approved):
I rebased your branch and changed directly some details.
Many thanks for your PR @platawiec !

platawiec commented:
> I rebased your branch and changed directly some details. Many thanks for your PR @platawiec !

No problem, happy to contribute! It seems like this is ready to go but I was wondering about your comment forcing dispatch to BLAS. If you don't mind, can you please post the benchmarks, since this is counter-intuitive to me?

amontoison commented:
Benchmarks for `scal!`:
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │       1.0 │  1.14368 │ 1.15129 │  1.0191 │
│     100.0 │   1.41832 │      1.0 │ 1.06149 │ 1.41976 │
│    1000.0 │   3.26601 │      1.0 │ 1.20754 │ 3.26612 │
│   10000.0 │    3.0177 │  1.29019 │     1.0 │ 3.01378 │
│  100000.0 │   3.24151 │  1.89749 │     1.0 │   3.342 │
│     1.0e6 │   1.30467 │  1.20134 │     1.0 │  1.3039 │
│     1.0e7 │   1.22009 │  1.08321 │     1.0 │ 1.22064 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```

Benchmarks for `axpy!`: 
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │       1.0 │  1.14882 │ 1.14913 │ 1.03988 │
│     100.0 │   1.50763 │      1.0 │ 1.01488 │ 1.81782 │
│    1000.0 │   2.84909 │      1.0 │ 1.01175 │ 4.22839 │
│   10000.0 │   3.23767 │  1.84615 │     1.0 │ 3.74639 │
│  100000.0 │   2.57948 │      1.0 │ 1.14925 │  2.4556 │
│     1.0e6 │   1.17641 │  1.05842 │     1.0 │ 1.22009 │
│     1.0e7 │   1.16617 │  1.07253 │     1.0 │ 1.25245 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```

Benchmarks for `axpby!`:
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │   1.00868 │  1.20266 │ 1.46403 │     1.0 │
│     100.0 │   1.26446 │  1.81398 │     1.0 │ 1.34513 │
│    1000.0 │   2.06898 │   3.0931 │     1.0 │  2.1006 │
│   10000.0 │   2.90107 │   4.2573 │     1.0 │ 2.86887 │
│  100000.0 │   2.30153 │  2.91977 │     1.0 │ 2.18438 │
│     1.0e6 │   1.21786 │  1.47258 │     1.0 │ 1.21689 │
│     1.0e7 │   1.20291 │  1.34659 │     1.0 │ 1.22869 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```

amontoison commented:
@platawiec
I also give you the code used for the benchmarks.
I have 2 physical cores on my laptop for information.

```
using BenchmarkTools, LinearAlgebra

@inline krylov_scal!(n :: Integer, s :: T, x :: Vector{T}, dx :: Integer) where T <: BLAS.BlasFloat = BLAS.scal!(n, s, x, dx)
@inline krylov_scal!(n :: Integer, s :: T, x :: AbstractVector{Complex{T}}, dx :: Integer) where T <: AbstractFloat = krylov_scal!(n, Complex{T}(s), x, dx)

@inline krylov_scal2!(n, s, x, dx) = (x .*= s)

function krylov_scal3!(n, s, x, dx)
  for IX in eachindex(x)
    @inbounds x[IX] *= s
  end
  return x
end

@inline krylov_axpy!(n :: Integer, s :: T, x :: Vector{T}, dx :: Integer, y :: Vector{T}, dy :: Integer) where T <: BLAS.BlasFloat = BLAS.axpy!(n, s, x, dx, y, dy)
@inline krylov_axpy!(n :: Integer, s :: T, x :: AbstractVector{Complex{T}}, dx :: Integer, y :: AbstractVector{Complex{T}}, dy :: Integer) where T <: AbstractFloat = krylov_axpy!(n, Complex{T}(s), x, dx, y, dy)

@inline krylov_axpy2!(n, s, x, dx, y, dy) = (y .+= s .* x)
@inline krylov_axpy3!(n, s, x, dx, y, dy) = axpy!(s, x, y)

function krylov_axpy3!(n, s, x, dx, y, dy)
  for (IY, IX) in zip(eachindex(y), eachindex(x))
    @inbounds y[IY] += x[IX]*s
  end
  return y
end

@inline krylov_axpby!(n :: Integer, s :: T, x :: Vector{T}, dx :: Integer, t :: T, y :: Vector{T}, dy :: Integer) where T <: BLAS.BlasFloat = BLAS.axpby!(n, s, x, dx, t, y, dy)
@inline krylov_axpby!(n :: Integer, s :: T, x :: AbstractVector{Complex{T}}, dx :: Integer, t :: Complex{T}, y :: AbstractVector{Complex{T}}, dy :: Integer) where T <: AbstractFloat = krylov_axpby!(n, Complex{T}(s), x, dx, t, y, dy)
@inline krylov_axpby!(n :: Integer, s :: Complex{T}, x :: AbstractVector{Complex{T}}, dx :: Integer, t :: T, y :: AbstractVector{Complex{T}}, dy :: Integer) where T <: AbstractFloat = krylov_axpby!(n, s, x, dx, Complex{T}(t), y, dy)
@inline krylov_axpby!(n :: Integer, s :: T, x :: AbstractVector{Complex{T}}, dx :: Integer, t :: T, y :: AbstractVector{Complex{T}}, dy :: Integer) where T <: AbstractFloat = krylov_axpby!(n, Complex{T}(s), x, dx, Complex{T}(t), y, dy)

@inline krylov_axpby2!(n, s, x, dx, t, y, dy) = (y .= s .* x .+ t.* y)

function krylov_axpby3!(n, s, x, dx, t, y, dy)
  for (IX, IY) in zip(eachindex(x), eachindex(y))
    @inbounds y[IY] = x[IX]*s + y[IY]*t
  end
  return y
end

p = 7
iter = 1000
benchmarks = false
FC = ComplexF64
T  = Float64

time_broadcast = zeros(3,p)
time_openblas  = zeros(3,p)
time_mkl       = zeros(3,p)
time_generic   = zeros(3,p)

println("--- Broadcast ---")
for i = 1:p
  n = 10^i
  println("Dimension: ", n)
  x = rand(FC, n)
  y = rand(FC, n)
  s = one(T) + rand(T)
  t = one(T) + rand(T)
  dx = 1
  dy = 1
  print("scal!  -- ")
  benchmarks && @btime for k in 1:iter krylov_scal2!($n, $s, $x, $dx) end
  time_broadcast[1,i] = @belapsed for k in 1:iter krylov_scal2!($n, $s, $x, $dx) end
  print("axpy!  -- ")
  benchmarks && @btime for k in 1:iter krylov_axpy2!($n, $s, $x, $dx, $y, $dy) end
  time_broadcast[2,i] = @belapsed for k in 1:iter krylov_axpy2!($n, $s, $x, $dx, $y, $dy) end
  print("axpby! -- ")
  benchmarks && @btime for k in 1:iter krylov_axpby2!($n, $s, $x, $dx, $t, $y, $dy) end
  time_broadcast[3,i] = @belapsed for k in 1:iter krylov_axpby2!($n, $s, $x, $dx, $t, $y, $dy) end
  println()
end

println("--- OpenBLAS ---")
for i = 1:p
  n = 10^i
  println("Dimension: ", n)
  x = rand(FC, n)
  y = rand(FC, n)
  s = one(T) + rand(T)
  t = one(T) + rand(T)
  dx = 1
  dy = 1
  print("scal!  -- ")
  benchmarks && @btime for k in 1:iter krylov_scal!($n, $s, $x, $dx) end
  time_openblas[1,i] = @belapsed for k in 1:iter krylov_scal!($n, $s, $x, $dx) end
  print("axpy!  -- ")
  benchmarks && @btime for k in 1:iter krylov_axpy!($n, $s, $x, $dx, $y, $dy) end
  time_openblas[2,i] = @belapsed for k in 1:iter krylov_axpy!($n, $s, $x, $dx, $y, $dy) end
  print("axpby! -- ")
  benchmarks && @btime for k in 1:iter krylov_axpby!($n, $s, $x, $dx, $t, $y, $dy) end
  time_openblas[3,i] = @belapsed for k in 1:iter krylov_axpby!($n, $s, $x, $dx, $t, $y, $dy) end
  println()
end

using MKL

println("--- MKL ---")
for i = 1:p
  n = 10^i
  println("Dimension: ", n)
  x = rand(FC, n)
  y = rand(FC, n)
  s = one(T) + rand(T)
  t = one(T) + rand(T)
  dx = 1
  dy = 1
  print("scal!  -- ")
  benchmarks && @btime for k in 1:iter krylov_scal!($n, $s, $x, $dx) end
  time_mkl[1,i] = @belapsed for k in 1:iter krylov_scal!($n, $s, $x, $dx) end
  print("axpy!  -- ")
  benchmarks && @btime for k in 1:iter krylov_axpy!($n, $s, $x, $dx, $y, $dy) end
  time_mkl[2,i] = @belapsed for k in 1:iter krylov_axpy!($n, $s, $x, $dx, $y, $dy) end
  print("axpby! -- ")
  benchmarks && @btime for k in 1:iter krylov_axpby!($n, $s, $x, $dx, $t, $y, $dy) end
  time_mkl[3,i] = @belapsed for k in 1:iter krylov_axpby!($n, $s, $x, $dx, $t, $y, $dy) end
  println()
end

println("--- Generic ---")
for i = 1:p
  n = 10^i
  println("Dimension: ", n)
  x = rand(FC, n)
  y = rand(FC, n)
  s = one(T) + rand(T)
  t = one(T) + rand(T)
  dx = 1
  dy = 1
  print("scal!  -- ")
  benchmarks && @btime for k in 1:iter krylov_scal3!($n, $s, $x, $dx) end
  time_generic[1,i] = @belapsed for k in 1:iter krylov_scal3!($n, $s, $x, $dx) end
  print("axpy!  -- ")
  benchmarks && @btime for k in 1:iter krylov_axpy3!($n, $s, $x, $dx, $y, $dy) end
  time_generic[2,i] = @belapsed for k in 1:iter krylov_axpy3!($n, $s, $x, $dx, $y, $dy) end
  print("axpby! -- ")
  benchmarks && @btime for k in 1:iter krylov_axpby3!($n, $s, $x, $dx, $t, $y, $dy) end
  time_generic[3,i] = @belapsed for k in 1:iter krylov_axpby3!($n, $s, $x, $dx, $t, $y, $dy) end
  println()
end

using PrettyTables

dimension  = [10^i for i=1:p]
time_scal  = hcat(dimension, time_broadcast[1,:], time_openblas[1,:], time_mkl[1,:], time_generic[1,:])
time_axpy  = hcat(dimension, time_broadcast[2,:], time_openblas[2,:], time_mkl[2,:], time_generic[2,:])
time_axpby = hcat(dimension, time_broadcast[3,:], time_openblas[3,:], time_mkl[3,:], time_generic[3,:])

pretty_table(time_scal , header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])
pretty_table(time_axpy , header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])
pretty_table(time_axpby, header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])

for i = 1:p
  val_scal  = minimum( time_scal[i,2:5])
  val_axpy  = minimum( time_axpy[i,2:5])
  val_axpby = minimum(time_axpby[i,2:5])
  for j = 2:5
    time_scal[i,j]  /= val_scal
    time_axpy[i,j]  /= val_axpy
    time_axpby[i,j] /= val_axpby
  end
end

pretty_table(time_scal , header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])
pretty_table(time_axpy , header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])
pretty_table(time_axpby, header = ["Dimension", "Broadcast", "OpenBLAS", "MKL", "Generic"])
```

amontoison commented:
Thanks @tmigot for also running the benchmarks !

Benchmarks for `scal!`:
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │       1.0 │  1.16615 │ 1.33846 │ 1.02462 │
│     100.0 │   1.32035 │   1.4329 │     1.0 │ 1.33333 │
│    1000.0 │   2.73626 │  2.44542 │     1.0 │ 2.73773 │
│   10000.0 │   3.68756 │  3.68893 │     1.0 │  3.6846 │
│  100000.0 │   8.97625 │  10.0532 │     1.0 │ 8.99271 │
│     1.0e6 │     3.113 │  3.79763 │     1.0 │ 2.97444 │
│     1.0e7 │   1.06403 │      1.0 │ 1.05432 │  1.0138 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```

Benchmarks for `axpy!`: 
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │       1.0 │  1.14943 │ 1.18103 │ 1.02874 │
│     100.0 │    1.5046 │  1.37569 │     1.0 │ 1.72007 │
│    1000.0 │   3.48685 │   2.6367 │     1.0 │ 3.42752 │
│   10000.0 │    4.4471 │  3.35508 │     1.0 │ 4.31941 │
│  100000.0 │   13.0104 │  7.64088 │     1.0 │  13.322 │
│     1.0e6 │   1.08753 │      1.0 │ 1.01735 │ 1.10467 │
│     1.0e7 │   1.39971 │      1.0 │ 1.57695 │ 1.38047 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```

Benchmarks for `axpby!`:
```
┌───────────┬───────────┬──────────┬─────────┬─────────┐
│ Dimension │ Broadcast │ OpenBLAS │     MKL │ Generic │
├───────────┼───────────┼──────────┼─────────┼─────────┤
│      10.0 │   1.00817 │  1.31608 │ 1.22343 │     1.0 │
│     100.0 │   1.37296 │  2.88762 │     1.0 │ 1.55537 │
│    1000.0 │   2.36191 │  5.83965 │     1.0 │ 2.58326 │
│   10000.0 │   3.94776 │  9.99935 │     1.0 │ 4.30499 │
│  100000.0 │   12.2748 │  30.7437 │     1.0 │ 13.4734 │
│     1.0e6 │   1.08952 │  1.43815 │     1.0 │ 1.10377 │
│     1.0e7 │   1.01062 │  1.04643 │ 1.12335 │     1.0 │
└───────────┴───────────┴──────────┴─────────┴─────────┘
```
```
julia> versioninfo()
Julia Version 1.7.0
Commit 3bf9d17731 (2021-11-30 12:12 UTC)
Platform Info:
OS: Windows (x86_64-w64-mingw32)
CPU: Intel(R) Xeon(R) W-1370 @ 2.90GHz
WORD_SIZE: 64
LIBM: libopenlibm
LLVM: libLLVM-12.0.1 (ORCJIT, icelake-client)
```

amontoison commented on test/test_utils.jl:
```diff
@@ -312,7 +328,7 @@ function small_sqd(transpose :: Bool=false; FC=Float64)
 end
 
-# Test restart feature with linear systems of size n³.
+# FCest restart feature with linear systems of size n³.
```
```suggestion
# Test restart feature with linear systems of size n³.
```

amontoison commented:
Thanks again @platawiec !
The new release 0.8.0 will have the support of complex numbers for all Krylov methods !
